postprocessing/science/spectral_corner_frequency_from_surface_xdmf.py

```diff
@@ -0,0 +1,387 @@
+    data2 = data[ind1:ind1+window_len] * taper
+    
+    result = np.zeros_like(data)
+    result[ind1:ind1+window_len] = data2
+    
+    return result
+
+def fc_Qzero_gridsearch(amplitude_tr, freq_tr, misfit="SSM", n=2, fc_end=0., q_dim=0):
+    """Determines the corner frequency of a displacement spectrum by performing a gridsearch 
+    to find the best fitting Brune-type spectrum"""
+    
+    misfit_types = ["RMS", "SSM"]
+    if misfit not in misfit_types:
+        raise ValueError("Invalid misfit type. Expected one of: %s" % misfit_types)
+    
+    # Prepare tensors for broadcasting
+    fc_ind = -2 if fc_end == 0. else int(np.argwhere(freq_tr>fc_end)[0])
+         
+    fc_tensor = np.empty((1,freq_tr[:fc_ind].shape[0], 1))
+    fc_tensor[0,:,0] = freq_tr[:fc_ind]
+    
+    freq_tensor = np.empty((freq_tr.shape[0],1,1))
+    freq_tensor[:,0,0] = freq_tr
+    
+    amplitude_tensor = np.empty((freq_tr.shape[0],1,1))
+    amplitude_tensor[:,0,0] = amplitude_tr
+    
+    # Initialise tensor of possible Q_zero values depending on amplitude_tr[0] or take mean amplitude below corner frequency
+    if q_dim != 0:
+        q_zero = np.array([[np.geomspace(0.1*amplitude_tr[0], 10*amplitude_tr[0], num=q_dim)]])
+    else:
+        q_zero = np.empty((1,freq_tr[:fc_ind].shape[0], 1))
+        for i in range(q_zero.shape[1]):
+            q_zero[0,i,0] = np.mean(amplitude_tr[:i+1])
+    
+    # Calculate Brune spectra of every (Q, fc) combination by broadcasting
+    brune_function = q_zero*(1/(1+(freq_tensor/fc_tensor)**n))
+    
+    # Calculate the misfit of every (Q, fc) combination
+    if misfit == "RMS":
+        misfit_tensor = (brune_function - amplitude_tensor)**2
+        misfit_tensor = np.sqrt(np.mean(misfit_tensor, axis=0))
+    if misfit == "SSM":
+        misfit_tensor = np.abs(np.log10(amplitude_tensor/brune_function))
+        misfit_tensor = np.mean(misfit_tensor, axis=0)
+    
+    # Get indices of minimum misfit (Q, fc) combination
+    index = np.unravel_index(np.argmin(misfit_tensor), misfit_tensor.shape)
+    
+    fc = fc_tensor[0,index[0],0]
+    if q_dim != 0:
+        q = q_zero[0,0,index[1]]
+    else:
+        q = q_zero[0,index[1],0]
+        
+    return fc, q
+
+def complete_fft(tr, dt):
+    tr_fft=np.fft.rfft(tr.data)
+    times_comp = np.fft.rfftfreq(tr.shape[0],dt)
+    return tr_fft, times_comp
+
+def ApproximateEventDurationAndHypocenter():
+    """Calculates the derivative of ASl (probably faster than loading SR1 and SR2)
+     and approximates event duration by multiplying dt with the number of timesteps,
+     where maximum on-fault slip rate is above a threshold (slipRateThreshold)"""
+    
+    slipRateThreshold=args.slipRateThreshold[0]
+    if nprocs == 1 or not args.parallelLoading:
+        ASl = faultxdmf.ReadData("ASl").T[::stepsize,timeIndicesFault[0]:timeIndicesFault[1]]
+    else:
+        ASl = LoadSingleComponentParallel("ASl", surfaceData=False)
+    ASR = np.gradient(ASl, dtFault, axis=1)
```

Comment:
I guess you need a np.abs also here?

Reply:
I added it, but I don't think it's necessary because ASl is monotonously increasing.

Reply:
ha right, sorry.